chain-extensions/pallet-assets/src/lib.rs

```diff
@@ -0,0 +1,411 @@
+use pallet_contracts::chain_extension::{
+    ChainExtension, Environment, Ext, InitState, RetVal, SysConfig, UncheckedFrom,
+};
+use sp_core::Get;
+use sp_runtime::traits::StaticLookup;
+use sp_runtime::DispatchError;
+use sp_std::marker::PhantomData;
+use sp_std::vec::Vec;
+
+enum AssetsFunc {
+    Create,
+    Transfer,
+    Mint,
+    Burn,
+    BalanceOf,
+    TotalSupply,
+    Allowance,
+    ApproveTransfer,
+    CancelApproval,
+    TransferApproved,
+    SetMetadata,
+    MetadataName,
+    MetadataSymbol,
+    MetadataDecimals,
+}
+
+impl TryFrom<u16> for AssetsFunc {
+    type Error = DispatchError;
+
+    fn try_from(value: u16) -> Result<Self, Self::Error> {
+        match value {
+            1 => Ok(AssetsFunc::Create),
+            2 => Ok(AssetsFunc::Transfer),
+            3 => Ok(AssetsFunc::Mint),
+            4 => Ok(AssetsFunc::Burn),
+            5 => Ok(AssetsFunc::BalanceOf),
+            6 => Ok(AssetsFunc::TotalSupply),
+            7 => Ok(AssetsFunc::Allowance),
+            8 => Ok(AssetsFunc::ApproveTransfer),
+            9 => Ok(AssetsFunc::CancelApproval),
+            10 => Ok(AssetsFunc::TransferApproved),
+            11 => Ok(AssetsFunc::SetMetadata),
+            12 => Ok(AssetsFunc::MetadataName),
+            13 => Ok(AssetsFunc::MetadataSymbol),
+            14 => Ok(AssetsFunc::MetadataDecimals),
+            _ => Err(DispatchError::Other(
+                "PalletAssetsExtension: Unimplemented func_id",
+            )),
+        }
+    }
+}
+
+/// Pallet Assets chain extension.
+pub struct AssetsExtension<T>(PhantomData<T>);
+
+impl<T> Default for AssetsExtension<T> {
+    fn default() -> Self {
+        AssetsExtension(PhantomData)
+    }
+}
+
+impl<T> ChainExtension<T> for AssetsExtension<T>
+where
+    T: pallet_assets::Config + pallet_contracts::Config,
+    <<T as SysConfig>::Lookup as StaticLookup>::Source: From<<T as SysConfig>::AccountId>,
+    <T as SysConfig>::AccountId: From<[u8; 32]>,
+{
+    fn call<E: Ext>(&mut self, env: Environment<E, InitState>) -> Result<RetVal, DispatchError>
+    where
+        E: Ext<T = T>,
+        <E::T as SysConfig>::AccountId: UncheckedFrom<<E::T as SysConfig>::Hash> + AsRef<[u8]>,
+    {
+        let func_id = env.func_id().try_into()?;
+        let mut env = env.buf_in_buf_out();
+
+        match func_id {
+            AssetsFunc::Create => {
+                let (origin, id, admin, min_balance): (
+                    Origin,
+                    <T as pallet_assets::Config>::AssetId,
+                    T::AccountId,
+                    T::Balance,
+                ) = env.read_as()?;
+
+                let base_weight = <T as pallet_assets::Config>::WeightInfo::create();
+                env.charge_weight(base_weight)?;
+
+                let runtime_origin = RawOrigin::Signed(match origin {
+                    Origin::Caller => env.ext().caller().clone(),
+                    Origin::Address => env.ext().address().clone(),
+                });
```

Comment:
Just to check - this isn't controlled directly by the caller but instead by context, right?

E.g. I cannot call this function as someone else :slightly_smiling_face:

Reply:
It's in the context. Chain-extension can only be called from contract. And in the contract call, the contract chose the origin 
https://github.com/AstarNetwork/chain-extension-contracts/blob/eb1b6892ef1ab002a8694f1f59678f0c138f2de3/examples/assets-mock/lib.rs#L39